### ISSUE TITLE
Using installed static library(with dependencies) with .pc file generated by pkgconfig module via plain dependency(·) call fails

### ISSUE THREAD

**Issue by textshell** [mesonbuild/meson#3074]
An installed static library with dependencies and an .pc file generated by the pkgconfig module is not usable with a plain `dependency(·)`, but only using `dependency(·, static: true)`. I think in the case that meson only installs a static library the user should not be forced to know to modify the dependency in an meson project that uses this library. This is especially true, when the library can be build in both variants.

I think meson should detect if `import('pkgconfig').generate` gets passed a static library and automatically promote it's dependencies from `Require.private` to `Require` (and similarly for `Libs`). This way meson could generate .pc files that work both with `-Ddefault_library=static` and `-Ddefault_library=shared`. For manually created 'both' passing the shared library to generate would still produce a .pc file that allows the using program to select static or dynamic linking.

Reproducer:

lib/liba.c:
```
void liba_func() {
}
```

lib/libb.c:
```
void liba_func();

void libb_func() {
    liba_func();
}
```

lib/meson.build
```
project('lib', ['c'])

a = library('static-test-a', 'liba.c', install: true)

b = library('static-test-b', 'libb.c', link_with: a, install: true)

import('pkgconfig').generate(
    version: '0.0',
    description: 'XXX',
    filebase: 'static-test-b',
    name: 'XXX',
    libraries: ['-Wl,-rpath-link,' + join_paths(get_option('prefix'), get_option('libdir')), b],
    subdirs: ['.']
)
```

app/app.c:
```
void libb_func();

int main() {
    libb_func();
    return 0;
}
```

app/meson.build:
```
project('app', ['c'])

b = dependency('static-test-b')

executable('app', 'app.c', dependencies : [b])
```

assuming a clean prefix:
```
cd lib
meson.py _build --prefix=$PWD/../prefix
meson.py configure _build -Ddefault_library=static
ninja -C _build/ install

cd ../app
PKG_CONFIG_PATH=$PWD/../prefix/lib/x86_64-linux-gnu/pkgconfig/ meson.py _build
ninja -C _build/
```

This work when omitting the `-Ddefault_library=static` line. But when building a static library it fails with:
```
FAILED: app 
ccache cc  -o app 'app@exe/app.c.o' -Wl,--no-undefined -Wl,--as-needed -L/home/foo/lib/../prefix/lib/x86_64-linux-gnu -Wl,-rpath-link,/home/foo/lib/../prefix/lib/x86_64-linux-gnu -Wl,--start-group -lstatic-test-b -Wl,--end-group  
/home/foo/lib/../prefix/lib/x86_64-linux-gnu/libstatic-test-b.a(libb.c.o): In function `libb_func':
/home/foo/lib/_build/../libb.c:4: undefined reference to `liba_func'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by xclaesse** [comment]
I don't think we should modify pc files, but this is related to https://github.com/mesonbuild/meson/pull/2816

**Comment by textshell** [comment]
This is not really related to #2816. I think a generated pkgconfig file should be usable without the consumer needed to know if it refers to a static or dynamic library (or both).
Even if both are installed the user should be able to get a working default without any further options(with both that would be dynamic linked).

Meson could get pc files that only have static libraries that are "unmodified" to work by first detecting that case (manually resolving to absolute pathes) and then automatically switching to pkg-configs static linking mode. But that would still leave other build systems unable to consume the generated pc files. Which seems like a bad situation. At least i think meson uses pc files to get basic interoperability with other build systems.

**Comment by textshell** [comment]
fixed via #3093
